quickwit-actors/src/actor_handle.rs

```diff
@@ -0,0 +1,165 @@
+/// to the actor type, `ActorHandle` are interchangeable.
+/// It makes it possible to plug different implementations, have actor proxy etc.use std::fmt;
+use tokio::sync::{oneshot, watch};
+use tokio::task::{JoinError, JoinHandle};
+use tokio::time::timeout;
+use tracing::error;
+
+pub struct ActorHandle<A: Actor> {
+    actor_context: ActorContext<A>,
+    last_state: watch::Receiver<A::ObservableState>,
+    join_handle: JoinHandle<ActorTermination>,
+}
+
+impl<A: Actor> fmt::Debug for ActorHandle<A> {
+    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+        formatter
+            .debug_struct("ActorHandle")
+            .field("name", &self.actor_context.actor_instance_name())
+            .finish()
+    }
+}
+
+impl<A: Actor> ActorHandle<A> {
+    pub(crate) fn new(
+        last_state: watch::Receiver<A::ObservableState>,
+        join_handle: JoinHandle<ActorTermination>,
+        ctx: ActorContext<A>,
+    ) -> Self {
+        let mut interval = tokio::time::interval(crate::HEARTBEAT);
+        let actor_instance_name = ctx.actor_instance_name().to_string();
+        let ctx_clone = ctx.clone();
+        tokio::task::spawn(async move {
+            interval.tick().await;
+            while ctx.kill_switch().is_alive() {
+                interval.tick().await;
+                if !ctx.progress().harvest_changes() {
+                    if ctx.get_state() == ActorState::Terminated {
+                        return;
+                    }
+                    error!(actor=%actor_instance_name, "actor-timeout");
+                    ctx.kill_switch().kill();
+                    return;
+                }
+            }
+        });
+        ActorHandle {
+            join_handle,
+            last_state,
+            actor_context: ctx_clone,
+        }
+    }
+
+    /// Process all of the pending message, and returns a snapshot of
+    /// the observable state of the actor after this.
+    ///
+    /// This method is mostly useful for tests.
+    ///
+    /// To actually observe the state of an actor for ops purpose,
+    /// prefer using the `.observe()` method.
+    ///
+    /// This method timeout if reaching the end of the message takes more than an HEARTBEAT.
+    /// Hence, it is only useful or unit tests.
+    pub async fn process_pending_and_observe(&self) -> Observation<A::ObservableState> {
+        let (tx, rx) = oneshot::channel();
+        if self
+            .actor_context
+            .mailbox()
+            .send_actor_message(ActorMessage::Observe(tx))
+            .await
+            .is_err()
+        {
+            error!("Failed to send message");
+        }
+        // TODO The timeout is required here. If the actor fails, the inbox is properly dropped but the send channel might actually
+        // prevent the onechannel Receiver from being dropped.
+        let observable_state_res = tokio::time::timeout(crate::HEARTBEAT, rx).await;
```

Comment:
I guess that's fine but as is but wouldn't it be nicer to have the `observable_state_res` contain the observed state? I find this API a bit clunky. Plus, what if in between the two calls, the state of the actor is modified and `observe` has been called somewhere else.

Reply:
I address this. It is a bit quirky though as I did not want to add ObservableState to the mailbox generic parameter.
I ended up relying on `Box<dyn Any>`. (The casting is done by the `ActorHandle` so everything is safe)